### PR TITLE
[Train] LightningTrainer support DeepSpeedStrategy

### LINKED ISSUE #35913: [Train] Support DeepSpeed Strategy in LightningTrainer
### Description

[DeepSpeed](https://github.com/microsoft/DeepSpeed) implements everything described in the [ZeRO paper](https://arxiv.org/abs/1910.02054). Currently it provides full support for:

- Optimizer state partitioning (ZeRO stage 1)
- Gradient partitioning (ZeRO stage 2)
- Parameter partitioning (ZeRO stage 3)
- Custom mixed precision training handling
- A range of fast CUDA-extension-based optimizers
- ZeRO-Offload to CPU and Disk/NVMe

LightningTrainer supports FSDP for distributed training, while DeepSpeed provides more fine-grained control over your distributed strategy. We need to integrate with current LightningTrainer API and together with an end-2-end example.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

Enabling DeepSpeedStrategy for LightningTrainer.

- Use distributed checkpointing for DeepSpeed strategy
- Change the working directory for all workers to `{trial_dir}/rank_all` to align with Lightning's setting.
- Tested auto-restoration under local storage setting

TODOs:
- End-to-end example for DeepSpeed Model Fine-tuning.
- Enable auto-restoration under cloud storage setting.

## Related issue number
Closes #35913, #35894

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/lightning/_lightning_utils.py [hunk 1]
```diff
@@ -194,17 +222,23 @@ def _session_report(self, trainer: "pl.Trainer", stage: str):
...
+            src_model_path = os.path.expanduser(self.last_model_path)
+            dst_model_path = f"{tmpdir}/{MODEL_KEY}"
+
+            if self.is_report_rank:
+                if os.path.isdir(src_model_path):
+                    shutil.copytree(src_model_path, dst_model_path)
+                elif os.path.isfile(src_model_path):
+                    shutil.copy(src_model_path, dst_model_path)
+
+            # Only the report_rank worker creates the actual checkpoints.
```
Comment (justinvyu) [inline python/ray/train/lightning/_lightning_utils.py:239]: Can you add some more description here to describe the behavior for ddp vs. deepspeed?

Something like:

```
For DDP: Each node has the same set of weights, so only the global rank 0 worker needs to report.
For deepspeed: Each node has a unique set of weights, so the local rank 0 workers report the checkpoint shards for all workers on its node. 
```

File: python/ray/train/lightning/lightning_trainer.py [hunk 2]
```diff
@@ -453,6 +456,13 @@ def restore(
 
 def _lightning_train_loop_per_worker(config):
     """Per-worker training loop for a Lightning Trainer."""
+
+    # Change the working directory for all workers to the same directory
+    # This aligns with Lightning's settings and avoids inconsistency for checkpointing.
```
Comment (justinvyu) [inline python/ray/train/lightning/lightning_trainer.py:461]: Add some examples of what could be inconsistent (model checkpoint state, etc.)

File: python/ray/train/lightning/_lightning_utils.py [hunk 3]
```diff
@@ -194,17 +222,23 @@ def _session_report(self, trainer: "pl.Trainer", stage: str):
...
         with tempfile.TemporaryDirectory() as tmpdir:
-            if trainer.global_rank == 0:
-                shutil.copy(self.last_model_path, f"{tmpdir}/{MODEL_KEY}")
-                checkpoint = LightningCheckpoint.from_directory(path=tmpdir)
-            else:
-                checkpoint = LightningCheckpoint.from_dict(
-                    {"rank": session.get_world_rank()}
-                )
+            src_model_path = os.path.expanduser(self.last_model_path)
+            dst_model_path = f"{tmpdir}/{MODEL_KEY}"
```
Comment (justinvyu) [inline python/ray/train/lightning/_lightning_utils.py:231]: Can use os.path.join

File: python/ray/train/tests/test_lightning_deepspeed.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    # The code after the yield will run as teardown code.
+    ray.shutdown()
+
+
+@pytest.mark.parametrize("stage", [1, 2, 3])
+@pytest.mark.parametrize("test_restore", [True, False])
+def test_deepspeed_stages(ray_start_6_cpus_4_gpus, tmpdir, stage, test_restore):
+    # Reduce tests number
+    if test_restore and stage < 3:
+        return
```
Comment (justinvyu) [inline python/ray/train/tests/test_lightning_deepspeed.py:27]: Restoring works for stage 3? Maybe modify this comment to be a TODO to test stage 3 + restore once it's possible.

File: python/ray/train/tests/test_lightning_deepspeed.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            ),
+        ),
+    )
+
+    if test_restore:
+        # TODO(yunxuanx): Remove this after we determined the error behavior
+        try:
+            trainer.fit()
+        except Exception:
+            pass
```
Comment (justinvyu) [inline python/ray/train/tests/test_lightning_deepspeed.py:72]: with pytest.raises(RuntimeError)

File: python/ray/train/data_parallel_trainer.py [hunk 6]
```diff
@@ -468,7 +468,12 @@ def clear_lazy_checkpoint_marker():
                 logger.debug(
                     f"Deleting the stale lazy checkpoint marker file: {marker_file}."
                 )
-                marker_file.unlink()
+                # Multiple workers on the same node may delete this file at the
+                # same time. Return if the marker file has been deleted.
+                try:
+                    marker_file.unlink()
```
Comment (justinvyu) [inline python/ray/train/data_parallel_trainer.py:474]: Oh oops, I just opened this PR since I saw a flaky test caused by this 😅  https://github.com/ray-project/ray/pull/36426

I think I prefer just having the local rank 0 worker call unlink, but also fine with this.

File: python/ray/train/tests/lightning_test_utils.py [hunk 7]
```diff
@@ -4,14 +4,17 @@
...
 
 class LinearModule(pl.LightningModule):
-    def __init__(self, input_dim, output_dim, strategy="ddp") -> None:
+    def __init__(self, input_dim, output_dim, strategy="ddp", restore_epoch=-1) -> None:
         super().__init__()
         self.linear = nn.Linear(input_dim, output_dim)
         self.loss = []
         self.strategy = strategy
+        self.restored = session.get_checkpoint() is not None
+        self.restore_epoch = restore_epoch
```
Comment (justinvyu) [inline python/ray/train/tests/lightning_test_utils.py:17]: shouldl this be `fail_epoch`?

**Review by justinvyu — COMMENTED** [review]
Left some comments, generally looking good:

File: python/ray/train/lightning/_lightning_utils.py (hunk 1 above)
Comment (woshiyyya) [reply python/ray/train/lightning/_lightning_utils.py:239]: ok

File: python/ray/train/tests/test_lightning_deepspeed.py (hunk 4 above)
Comment (woshiyyya) [reply python/ray/train/tests/test_lightning_deepspeed.py:27]: I sort of forgot why it didn't work for stage 3. 
Just to clear our minds. It works for driver storage, but not for cloud storage. It's because each worker won't download checkpoint from the cloud on restore?

File: python/ray/train/tests/test_lightning_deepspeed.py (hunk 5 above)
Comment (woshiyyya) [reply python/ray/train/tests/test_lightning_deepspeed.py:72]: OK.

File: python/ray/train/tests/test_lightning_deepspeed.py (hunk 4 above)
Comment (justinvyu) [reply python/ray/train/tests/test_lightning_deepspeed.py:27]: Yeah, there's some issues with downloading the latest checkpoint from cloud if a local copy exists.

Comment (justinvyu) [reply python/ray/train/tests/test_lightning_deepspeed.py:27]: https://github.com/ray-project/ray/issues/36357

**Review by justinvyu — APPROVED** [review]
Thanks, lgtm. @Yard1 can you take a look before merging?

File: python/ray/train/lightning/_lightning_utils.py [hunk 8]
```diff
@@ -172,10 +188,25 @@ class RayModelCheckpoint(ModelCheckpoint):
...
+        super().setup(trainer, pl_module, stage)
         self.is_checkpoint_step = False
 
+        if isinstance(trainer.strategy, DeepSpeedStrategy):
+            # For DeepSpeed, each node has a unique set of param and optimizer states,
+            # so the local rank 0 workers report the checkpoint shards for all workers
+            # on their node.
+            self.is_report_rank = session.get_local_rank() == 0
+        else:
+            # For DDP and FSDP, only the global rank 0 worker saves the full model.
```
Comment (Yard1) [inline python/ray/train/lightning/_lightning_utils.py:206]: is that true for FSDP in all cases? I thought it depends on a setting

Comment (woshiyyya) [reply python/ray/train/lightning/_lightning_utils.py:206]: Yes, one can switch to save one shard per workerfor FSDP. However, currently Lightning by default gathers all shards to Rank 0. We can update this logic after they support it.

File: python/ray/train/tests/test_lightning_deepspeed.py (hunk 4 above)
Comment (woshiyyya) [reply python/ray/train/tests/test_lightning_deepspeed.py:27]: Ok. Got it!

**Review by Yard1 — APPROVED** [review]
